pkg/observability/config.go

```diff
@@ -0,0 +1,143 @@
+
+	// ReqLogTemplateKey is the CM key for the request log template.
+	RequestLogTemplateKey = "logging.request-log-template"
+
+	// EnableReqLogKey is the CM key to enable request log.
+	EnableRequestLogKey = "logging.enable-request-log"
+
+	// EnableProbeReqLogKey is the CM key to enable request logs for probe requests.
+	EnableProbeRequestLogKey = "logging.enable-probe-request-log"
+)
+
+type (
+	BaseConfig    = pkgo11y.Config
+	MetricsConfig = pkgo11y.MetricsConfig
+	RuntimeConfig = pkgo11y.RuntimeConfig
+	TracingConfig = pkgo11y.TracingConfig
+)
+
+// +k8s:deepcopy-gen=true
+type Config struct {
+	BaseConfig
+
+	RequestMetrics MetricsConfig `json:"requestMetrics"`
+
+	// EnableVarLogCollection specifies whether the logs under /var/log/ should be available
+	// for collection on the host node by the fluentd daemon set.
+	EnableVarLogCollection bool `json:",omitempty"`
+
+	// LoggingURLTemplate is a string containing the logging url template where
+	// the variable REVISION_UID will be replaced with the created revision's UID.
+	LoggingURLTemplate string `json:",omitempty"`
+
+	// EnableRequestLog enables activator/queue-proxy to write request logs.
+	EnableRequestLog bool `json:"enableRequestLog,omitempty"`
+
+	// RequestLogTemplate is the go template to use to shape the request logs.
+	RequestLogTemplate string `json:"requestLogTemplate,omitempty"`
+
+	// EnableProbeRequestLog enables queue-proxy to write health check probe request logs.
+	EnableProbeRequestLog bool `json:"enableProbeRequestLog,omitempty"`
+}
+
+func (c *Config) Validate() error {
+	if c.RequestLogTemplate == "" && c.EnableRequestLog {
+		return fmt.Errorf("%q was set to true, but no %q was specified", EnableRequestLogKey, RequestLogTemplateKey)
+	}
+
+	if c.RequestLogTemplate != "" {
+		// Verify that we get valid templates.
+		if _, err := texttemplate.New("requestLog").Parse(c.RequestLogTemplate); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func DefaultConfig() *Config {
+	return &Config{
+		BaseConfig:         *pkgo11y.DefaultConfig(),
+		RequestMetrics:     metrics.DefaultConfig(),
+		LoggingURLTemplate: DefaultLogURLTemplate,
+		RequestLogTemplate: DefaultRequestLogTemplate,
+	}
+}
+
+func NewFromMap(m map[string]string) (*Config, error) {
+	c := DefaultConfig()
+
+	if cfg, err := pkgo11y.NewFromMap(m); err != nil {
+		return nil, err
+	} else {
+		c.BaseConfig = *cfg
+	}
+
+	if rm, err := metrics.NewFromMapWithPrefix("request-", m); err != nil {
+		return nil, err
+	} else {
+		c.RequestMetrics = rm
+	}
+
+	err := configmap.Parse(m,
+		configmap.As("logging.enable-var-log-collection", &c.EnableVarLogCollection),
+		configmap.As("logging.revision-url-template", &c.LoggingURLTemplate),
```

Comment:
can we refactor these keys out into constants?

Reply:
I'd like to skip this for now. Long term I want to move these keys to `config-logging`

Reply:
that sounds reasonable